### PR TITLE
Optimize int precision

### LINKED ISSUE #1825: Datasets library not suitable for huge text datasets.
Hi,

I'm trying to use datasets library to load a 187GB dataset of pure text, with the intention of building a Language Model. The problem is that from the 187GB it goes to some TB when processed by Datasets. First of all, I think the pre-tokenizing step (with tokenizer.map()) is not really thought for datasets this big, but for fine-tuning datasets, as this process alone takes so much time, usually in expensive machines (due to the need of tpus - gpus) which is not being used for training. It would possibly be more efficient in such cases to tokenize each batch at training time (receive batch - tokenize batch - train with batch), so that the whole time the machine is up it's being used for training. 
Moreover, the pyarrow objects created from a 187 GB datasets are huge, I mean, we always receive OOM, or No Space left on device errors when only 10-12% of the dataset has been processed, and only that part occupies 2.1TB in disk, which is so many times the disk  usage of the pure text (and this doesn't make sense, as tokenized texts should be lighter than pure texts).

Any suggestions??

### LINKED ISSUE #1973: Question: what gets stored in the datasets cache and why is it so huge?
I'm running several training jobs (around 10) with a relatively large dataset (3M samples). The datasets cache reached 178G and it seems really large. What is it stored in there and why is it so large? I don't think I noticed this problem before and seems to be related to the new version of the datasets library. Any insight? Thank you!

### PR DESCRIPTION
**Pull request by albertvillanova**
Optimize int precision to reduce dataset file size.

Close #1973, close #1825

### CODE REVIEW

**Comment by albertvillanova** [comment]
@lhoestq, are the tests OK? Some other cases I missed? Do you agree with this approach?

**Review by lhoestq — COMMENTED** [review]
Cool !

File: src/datasets/arrow_writer.py [hunk 1]
```diff
@@ -123,6 +129,18 @@ def __arrow_array__(self, type=None):
...
 
+class OptimizedTypedSequence(TypedSequence):
+    def __init__(self, data, type=None, try_type=None, col=None, optimized_int_type=pa.int32()):
+        if type is None and try_type is None:
+            if col == "attention_mask":  # binary tensor
+                # pa.bool_()  # ArrowInvalid: Could not convert 1 with type int: tried to convert to boolean
+                optimized_int_type = pa.int8()
+            elif col == "token_type_ids":
+                # binary mask; Some models, like XLNetModel use an additional token represented by a 2
+                optimized_int_type = pa.int8()
```
[the remark this answers is not in the document]
Comment (albertvillanova) [reply src/datasets/arrow_writer.py:144]: Done!

File: src/datasets/arrow_writer.py [hunk 2]
```diff
@@ -123,6 +129,18 @@ def __arrow_array__(self, type=None):
                 raise
 
 
+class OptimizedTypedSequence(TypedSequence):
+    def __init__(self, data, type=None, try_type=None, col=None, optimized_int_type=pa.int32()):
```
[the remark this answers is not in the document]
Comment (albertvillanova) [reply src/datasets/arrow_writer.py]: I have fixed this: now default int is int64, and for `token_ids` is int32.

File: src/datasets/arrow_writer.py [hunk 3]
```diff
@@ -99,6 +100,11 @@ def __arrow_array__(self, type=None):
                 raise TypeError(
                     "Specified try_type alters data. Please check that the type/feature that you provided match the type/features of the data."
                 )
+            if self.optimized_int_type and self.type is None and self.try_type is None:
+                if pa.types.is_int64(out.type):
+                    out = out.cast(self.optimized_int_type)
+                elif pa.types.is_list(out.type) and pa.types.is_int64(out.type.value_type):
+                    out = out.cast(pa.list_(self.optimized_int_type))
```
Comment (albertvillanova) [inline src/datasets/arrow_writer.py:107]: @lhoestq might there be other cases besides scalar int and list of int?

Comment (lhoestq) [reply src/datasets/arrow_writer.py:107]: Maybe add list of list of int but honestly this is an extremely rare case

**Review by lhoestq — COMMENTED** [review]
Cool !

File: src/datasets/arrow_writer.py (hunk 3 above)
Comment (albertvillanova) [reply src/datasets/arrow_writer.py]: Done.

File: src/datasets/arrow_writer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (thomwolf) [reply src/datasets/arrow_writer.py]: I think int should default to `int32` no ?

[the remark this answers is not in the document]
Comment (albertvillanova) [reply src/datasets/arrow_writer.py]: As you prefer, @lhoestq and @thomwolf ... The key question is: what are the int range values in datasets? Can we use at most `int32` precision in any case? If so, I can set `int32` as default.

[the remark this answers is not in the document]
Comment (albertvillanova) [reply src/datasets/arrow_writer.py]: If not, then we could leave default `int64`, and specify individually `int32` by name...

[the remark this answers is not in the document]
Comment (lhoestq) [reply src/datasets/arrow_writer.py]: The general default should stay int64 IMO for consistency with numpy/torch/tf/pandas/arrow.
But for the tokens it should be int32 to save space.

Moreover it's actually not `token_ids` but `input_ids`, can you change that ?

**Comment by lhoestq** [comment]
I just tested this and it works like a charm :) 

However tokenizing and then setting the format to "torch" to feed the tokens into a model doesn't seem to work anymore, since the pytorch tensors have the int32/int8 precisions instead of int64 that is required as model inputs.

For example:

```python
import torch
from datasets import Dataset
from transformers import BertModel, BertTokenizer

torch.set_grad_enabled(False)

tokenizer = BertTokenizer.from_pretrained("bert-base-uncased")
model = BertModel.from_pretrained("bert-base-uncased")

dataset = Dataset.from_dict({"text": ["hello there !"]})
dataset = dataset.map(tokenizer, input_columns="text", remove_columns=dataset.column_names)
dataset = dataset.with_format("torch")

print(dataset.features)
# {'attention_mask': Sequence(feature=Value(dtype='int8', id=None), length=-1, id=None),
#  'input_ids': Sequence(feature=Value(dtype='int64', id=None), length=-1, id=None),  # this should be int32 though
#  'token_type_ids': Sequence(feature=Value(dtype='int8', id=None), length=-1, id=None)}

model(**dataset[:1])
# RuntimeError: Expected tensor for argument #1 'indices' to have scalar type Long; but got torch.CharTensor instead (while checking arguments for embedding)

dataset = dataset.with_format("torch", dtype=torch.int64)

model(**dataset[:1])
# works as expected
```

Pinging @sgugger here to make sure we take the right decision here.

Do we want the "torch" format to always return int64 ? Or does it have to keep the precision defined by the `dataset.features` 
 and therefore we would need to specify "torch" with `dtype=torch/int64` ?

File: src/datasets/arrow_writer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (albertvillanova) [reply src/datasets/arrow_writer.py]: @lhoestq Should I change:
- `token_ids` -> `input_ids`
- `token_type_ids` -> `input_type_ids`
- `special_tokens_mask` -> `special_ids_mask`
?!?

**Comment by sgugger** [comment]
From a user perspective, I think it's fine if the "torch" format converts all ints types to `torch.int64` by default since it's what the model will need almost all the time. I don't see a case where you would want to keep the low precision at the top of my head, and one can always write a custom transform for an edge case.

**Comment by lhoestq** [comment]
Sounds good to me !
For consistency maybe we should make the float precision fixed as well (float32, I guess)

**Comment by sgugger** [comment]
Yes, that would be the one used by default.

**Comment by albertvillanova** [comment]
Do we have the same requirements for TensorFlow?

File: src/datasets/arrow_writer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (lhoestq) [reply src/datasets/arrow_writer.py]: The final names should be
- input_ids
- token_type_ids
- special_tokens_mask
- attention_mask

So you just need to rename token_ids -> input_ids

[the remark this answers is not in the document]
Comment (albertvillanova) [reply src/datasets/arrow_writer.py]: Yes, it is done.

**Comment by lhoestq** [comment]
Yes I we should do the same for tensorflow as well since tf models would have the same issue

Thanks for adding this :)

File: src/datasets/arrow_writer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (lhoestq) [reply src/datasets/arrow_writer.py]: Thanksss :D

**Comment by albertvillanova** [comment]
@lhoestq I think this PR is ready... :)

**Review by lhoestq — APPROVED** [review]
LGTM

Thank you :)

**Pull request merged**